### ISSUE TITLE
multi-gpu support

### ISSUE THREAD

**Issue by Josh-Joseph** [mit-quest/necstlab-damage-segmentation#35]

**Comment by rak5216** [comment]
**### training speed comparison** with `configs/esrf16_segV1_study1/train-esrf16_segV1_study1_CropCase2_TrnValCaseA.yaml` (note val from `dataset-small-3class`):

1. Baseline Case: 1x P100 GPU, 4x CPU, batch size = 16 (72 training steps)
- epoch time of 70 - 90 s

2. Multi GPU Case I: 2x P100 GPU, 4x CPU, batch size = 16 (72 training steps)
- epoch time of 47 s
![image](https://user-images.githubusercontent.com/55461758/75406452-7bba1c80-58de-11ea-942b-8f77f67dae5a.png)

3. Multi GPU Case II: 2x P100 GPU, 4x CPU, batch size = 32 (36 training steps)
- epoch time of 38 s
![image](https://user-images.githubusercontent.com/55461758/75406327-30077300-58de-11ea-89c7-6e9d97fb0d13.png)

**Comment by rak5216** [comment]
@Josh-Joseph  can you comment on weird metrics warning image under Multi GPU Case III?

**Comment by Josh-Joseph** [comment]
So the warnings here aren't something to worry about in the sense of an error/value computed correctly. This is just a warning that a vector being passed to `update_state` is a python object (eg numpy array) and not a tensorflow tensor. A result of that is tensorflow has to rebuild the computation graph every time it's called (which is likely unneeded and can be expensive).

so with the current performances of 
- 1 GPU (case 1) 80s
- 2 GPUs (case 3) 38s
- 4 GPUs (case 5) 25s
I'm not *too* worried about it since the 4 GPU case is still a pretty significant speed up and it's sort of hard to tell exactly the improvement with such short epochs.

I think this will take some digging to figure out how best to resolve it so imo the best move here is to PR/merge this in to use for the experiments and create a new issue to resolve this. Seeing the performance of 4 or 8 GPUs on longer runs will give us more information on how important or not that new issue will be resolve.

**Comment by rak5216** [comment]
https://stackoverflow.com/questions/58972225/tensorflow2-warning-using-tffunction

**Comment by Josh-Joseph** [comment]
looks right as the thing we need to fix -- I think our question though is which vectors/where in the code is causing this issue?

**Comment by rak5216** [comment]
I tried disabling eager mode in tf, but didn't help as I ran into a new issue -- "AssertionError"

**Comment by rak5216** [comment]
it seems to be only related to the one hot metrics and the custom global metrics

**Comment by rak5216** [comment]
https://stackoverflow.com/questions/58814130/tensorflow-2-0-custom-keras-metric-caused-tf-function-retracing-warning

**Comment by rak5216** [comment]
https://www.tensorflow.org/tutorials/distribute/keras#define_the_callbacks

**Comment by rak5216** [comment]
https://blog.pythian.com/distributed-training-keras-models/

**Comment by rak5216** [comment]
https://stackoverflow.com/questions/59185729/how-to-use-keras-utils-sequence-data-generator-with-tf-distribute-mirroredstrate

**Comment by rak5216** [comment]
https://stackoverflow.com/questions/55852831/tf-data-vs-keras-utils-sequence-performance

**Comment by rak5216** [comment]
https://github.com/keras-team/keras/issues/12847

**Comment by CarolinaFurtado** [comment]
enable gpu monitoring: 
https://medium.com/@devops_83824/enabling-custom-gpu-metrics-in-stackdriver-for-google-cloud-platform-78067af32b92
https://cloud.google.com/compute/docs/gpus/monitor-gpus#setup-script

**Comment by CarolinaFurtado** [comment]
![image](https://user-images.githubusercontent.com/43469359/99583439-814a1c00-29b1-11eb-8eb9-d179bd9649fd.png)

**Comment by CarolinaFurtado** [comment]
We need to prefetch the data so it is prepared to feed the multiple gpus: 
https://keras.io/guides/distributed_training/

This is easy to do when the training data is a tf.data (A tf.data dataset. Should return a tuple of either (inputs, targets) or (inputs, targets, sample_weights) ) : https://towardsdatascience.com/how-to-reduce-training-time-for-a-deep-learning-model-using-tf-data-43e1989d2961
```
train_dataset =train_dataset.prefetch(tf.data.experimental.AUTOTUNE )
val_dataset =val_dataset.prefetch(tf.data.experimental.AUTOTUNE )
```


but we are using a generator: A generator or keras.utils.Sequence returning (inputs, targets) or (inputs, targets, sample_weights). A more detailed description of unpacking behavior for iterator types (Dataset, generator, Sequence).

so we need to find a way to: 
- or use tf.data:  
- or find a way to prefetch the generator data

**Comment by CarolinaFurtado** [comment]
**1. on the generator -** 

https://stanford.edu/~shervine/blog/keras-how-to-generate-data-on-the-fly#keras-script:~:text=Note%20that%20our%20implementation%20enables%20the,the%20GPU%20(and%20not%20data%20generation).

https://www.tensorflow.org/api_docs/python/tf/keras/Model

# Train model on dataset
model.fit_generator(generator=training_generator,
                    validation_data=validation_generator,
                    use_multiprocessing=True,
                    workers=6)

Note that our implementation enables the use of the multiprocessing argument of fit_generator, where the number of threads specified in workers are those that generate batches in parallel. A high enough number of workers assures that CPU computations are efficiently managed, i.e. that the bottleneck is indeed the neural network's forward and backward operations on the GPU (and not data generation).

**2. run longer epochs and see the gpu usage plot**

**3. add cpu usage to the plots**

**Comment by CarolinaFurtado** [comment]
## should we convert the generator to tf.data? 
https://sknadig.dev/TensorFlow2.0-dataset/

**Comment by CarolinaFurtado** [comment]
more helpful links: 
https://www.sicara.ai/blog/tensorflow-tutorial-training-time

**Comment by CarolinaFurtado** [comment]
![image](https://user-images.githubusercontent.com/43469359/106762736-21543080-6604-11eb-9d1a-e38ba2bbc9bb.png)

**Comment by CarolinaFurtado** [comment]
Are the results repeatable using tf data vs generator sequence? 
both run in cpu 
![image](https://user-images.githubusercontent.com/43469359/106762909-519bcf00-6604-11eb-82fd-a43e375f1cba.png)
